server/viame_server/viame_detection.py

```diff
@@ -274,16 +276,31 @@ def get_clip_meta(self, folder):
             required=True,
             level=AccessType.READ,
         )
-        .jsonParam("tracks", "", paramType="body")
+        .jsonParam(
+            "tracks",
+            "upsert and delete tracks",
+            paramType="body",
+            requireObject=True)
     )
     def save_detection(self, folder, tracks):
         user = self.getCurrentUser()
-        saveTracks(folder, tracks, user)
-        # TODO: verify schema before save.
-        # Right now the data object is too large in many cases and this takes too long to complete.
-        # dataclass_tracks = {
-        #     # Config kwarg needed to convert lists into tuples
-        #     trackId: from_dict(models.Track, track, config=Config(cast=[Tuple]))
-        #     for trackId, track in tracks.items()
-        # }
-        return True
+        upsert: Dict[str, dict] = tracks.get('upsert', {})
+        delete: List[str] = tracks.get('delete', [])
+        track_dict = getTrackData(self._load_detections(folder))
+
+        for track_id in delete:
+            del track_dict[str(track_id)]
+        for track_id, track in upsert.items():
+            validated: models.Track = from_dict(models.Track, track, config=Config(cast=[Tuple]))
+            track_dict[str(track_id)] = validated.asdict()
```

Comment:
Relates to my other comment in which your delete and upsert can contain the same track.  This is especially a problem if you try to delete a track before you upsert it.  See my other comment.  This could be modified by swapping the order but then an inefficiency is caused.  Probably better to handle it client side and remove items from upserting that are deleted?

Reply:
> This could be modified by swapping the order but then an inefficiency is caused

That could introduce another edge case.

* create track 1 and save, and refresh page
* delete track 1, and create a new track which gets allocated id 1, and save.

If you swap order, you'll save an empty file no matter what you do on the client.  You have to delete first.

I've replaced the `del dict[key]` line with `track_dict.pop(str(track_id), None)` so that delete fails silently if no key exists.

Reply:
Resolved in 349c5c5